master/internal/usergroup/api_groups.go

```diff
@@ -125,6 +161,18 @@ func (a *UserGroupAPIServer) UpdateGroup(ctx context.Context, req *apiv1.UpdateG
 		err = apiutils.MapAndFilterErrors(err)
 	}()
 
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	canUpdate, err := AuthZProvider.Get().CanUpdateGroups(*curUser)
```

Comment:
```CanUpdateGroup``` here (and in delete group) doesn't need to return a boolean since this won't cause any more information about what groups exist to be gotten

If we are going the route that it is all or nothing, you can update all groups or no groups then 
```
if err := AuthZProvider.Get().CanUpdateGroups(*curUser); err != nil {
    return status.Error(codes.PermissionDenied, err.Error())
}
```
is good

If we wanted to have ```CanUpdateGroups``` be conditional on the group then we would want to do

```
if canSee, serverErr = AuthZProvider.Get().CanGetGroup(*curUser, gid); serverErr != nil {
    return serverError
} else if !canSee {
    return errors.Wrapf(db.ErrNotFound, "Error getting group %d", gid) // Same error as when group actually doesn't 
}  
if err := AuthZProvider.Get().CanUpdateGroups(*curUser, gid); err != nil {
    return status.Error(codes.PermissionDenied, err.Error())
}

```

Reply:
We use `CanUpdateGroup` in group creation, so I think we'll need the boolean for server vs access denied error.

Reply:
Perhaps I'm missing something, but we could just declare an unexported sentinel variable
```go
var errorAccessDenied = fmt.Errorf("access denied")
```
then calling code could check
```go
err := CanUpdateGroups(...)
if err == errorAccessDenied {
	return nil, status.Errorf(codes.NotFound, "group not found: %d", req.GroupId)
} else if err != nil {
	return nil, err
}
```

I agree that it seems awkward for the API of the interface to be to return a `bool` and an `error` if only one of the two is needed.

Reply:
As for returning a sentinel vs boolean in something like ```CanGetGroup``` the boolean and ```serverError``` in my opinion makes the interface easier to implement correctly and harder to misuse. Requiring calling code to check the error returned makes it pretty easy to forgot the sentinel check and just do
```
if err := CanGetGroup(); err != nil {
    return err
}
```

While forgetting to check the boolean version becomes a compiler error
```
if canGetGroup, err := CanGetGroup(); err != nil {
    return err
}
```

Reply:
After syncing with Nick, the consensus is to essentially do sentinel errors - we'll modify the returned error from database and we'll edit the `mapAndFilterErrors` function to pass that error through.

Reply:
the `mapAndFilterErrors` change will be pulled in with the authz implementation PR for user groups since that function is only in EE